File: frontend/src/lib/schemas/memory.ts
/** Canonical SnapAct types. These mirror the Postgres enums exactly. */

export const CONTENT_TYPES = [
  "event",
  "place",
  "product",
  "person",
  "job",
  "quote",
  "knowledge",
  "idea",
  "task",
  "message",
  "media",
  "document",
  "receipt",
  "app_ui",
  "other",
] as const;
export type ContentType = (typeof CONTENT_TYPES)[number];

export const INTENT_MODES = ["REMEMBER", "EXPLORE", "ACT"] as const;
export type IntentMode = (typeof INTENT_MODES)[number];

export const URGENCIES = ["none", "low", "medium", "high"] as const;
export type Urgency = (typeof URGENCIES)[number];

export const MEMORY_STATUSES = ["pending", "ready", "failed"] as const;
export type MemoryStatus = (typeof MEMORY_STATUSES)[number];

export const ACTION_TYPES = [
  "open_url",
  "register",
  "add_calendar",
  "remind",
  "follow_up",
  "research",
  "buy",
  "view",
  "none",
] as const;
export type ActionType = (typeof ACTION_TYPES)[number];

export type CaptureMode = "save" | "ask" | "describe";

/** Human-facing labels, used by the UI and by prompts so both agree. */
export const CONTENT_TYPE_LABELS: Record<ContentType, string> = {
  event: "Event",
  place: "Place",
  product: "Product",
  person: "Person",
  job: "Job",
  quote: "Quote",
  knowledge: "Knowledge",
  idea: "Idea",
  task: "Task",
  message: "Conversation",
  media: "Media",
  document: "Document",
  receipt: "Receipt",
  app_ui: "App screen",
  other: "Other",
};

export interface Entity {
  name: string;
  type: string;
}

export interface SuggestedAction {
  type: ActionType | string;
  label: string;
  url?: string | null;
  due_on?: string | null;
  reason?: string | null;
}

export interface Citation {
  title?: string | null;
  url: string;
  source?: string | null;
  snippet?: string | null;
}

/** What the vision model is asked to produce for one screenshot. */
export interface ScreenshotAnalysis {
  title: string;
  content_type: ContentType;
  intent_mode: IntentMode;
  intent_summary: string;
  description: string;
  ocr_text: string;
  tags: string[];
  entities: Entity[];
  actionable: boolean;
  urgency: Urgency;
  confidence: number;
  due_on: string | null;
  event_on: string | null;
  suggested_actions: SuggestedAction[];
  event: Record<string, unknown> | null;
  place: Record<string, unknown> | null;
  person: Record<string, unknown> | null;
  product: Record<string, unknown> | null;
  answer: string | null;
  citations: Citation[];
  short_message: string;
  agent_activity: string[];
}

/** One row of public.memories, as the app sees it. */
export interface Memory {
  id: string;
  user_id: string;
  client_request_id: string;
  status: MemoryStatus;

  source: string | null;
  capture_mode: string | null;
  captured_at: string | null;
  created_at: string;
  updated_at: string;

  image_path: string | null;
  image_bytes: number | null;
  image_mime: string | null;

  title: string | null;
  content_type: ContentType;
  intent_mode: IntentMode;
  intent_summary: string | null;
  description: string | null;
  ocr_text: string | null;
  tags: string[];
  entities: Entity[];

  actionable: boolean;
  urgency: Urgency;
  due_on: string | null;
  event_on: string | null;
  completed_at: string | null;
  suggested_actions: SuggestedAction[];

  event: Record<string, unknown> | null;
  place: Record<string, unknown> | null;
  person: Record<string, unknown> | null;
  product: Record<string, unknown> | null;

  user_note: string | null;
  user_question: string | null;
  answer: string | null;
  citations: Citation[];

  confidence: number | null;
  model: string | null;
  analysis: ScreenshotAnalysis | null;
  analysis_error: string | null;
  analysis_attempts: number;

  search_text: string | null;
}

/** A memory decorated with retrieval provenance. */
export interface RetrievedMemory extends Memory {
  image_url: string | null;
  similarity: number;
  lexical_rank: number;
  score: number;
  matched_by: "hybrid" | "lexical" | "semantic";
  /** Set once the relevance gate has judged this candidate. */
  relevance?: "primary" | "supporting";
  relevance_reason?: string | null;
}

export function isContentType(value: unknown): value is ContentType {
  return CONTENT_TYPES.includes(value as ContentType);
}

export function coerceContentType(value: unknown): ContentType {
  const raw = String(value ?? "").toLowerCase().trim();
  if (isContentType(raw)) return raw;
  // Tolerate near-misses from the model rather than silently dumping to "other".
  const aliases: Record<string, ContentType> = {
    person_followup: "person",
    people: "person",
    contact: "person",
    conversation: "message",
    tweet: "message",
    post: "message",
    dm: "message",
    chat: "message",
    entertainment: "media",
    movie: "media",
    film: "media",
    show: "media",
    tv: "media",
    music: "media",
    song: "media",
    book: "media",
    video: "media",
    restaurant: "place",
    venue: "place",
    location: "place",
    shopping: "product",
    item: "product",
    listing: "product",
    article: "knowledge",
    docs: "knowledge",
    documentation: "knowledge",
    reference: "knowledge",
    tutorial: "knowledge",
    note: "idea",
    thought: "idea",
    todo: "task",
    reminder: "task",
    ticket: "document",
    confirmation: "document",
    form: "document",
    invoice: "receipt",
    order: "receipt",
    purchase: "receipt",
    settings: "app_ui",
    home_screen: "app_ui",
    homescreen: "app_ui",
    ui: "app_ui",
    screen: "app_ui",
    app: "app_ui",
  };
  return aliases[raw] ?? "other";
}

export function coerceIntentMode(value: unknown): IntentMode {
  const raw = String(value ?? "").toUpperCase().trim();
  return (INTENT_MODES as readonly string[]).includes(raw) ? (raw as IntentMode) : "REMEMBER";
}

export function coerceUrgency(value: unknown): Urgency {
  const raw = String(value ?? "").toLowerCase().trim();
  return (URGENCIES as readonly string[]).includes(raw) ? (raw as Urgency) : "none";
}

/** Accepts YYYY-MM-DD only; anything else becomes null rather than a bogus date. */
export function coerceDate(value: unknown): string | null {
  const raw = String(value ?? "").trim();
  if (!/^\d{4}-\d{2}-\d{2}$/.test(raw)) return null;
  const parsed = new Date(`${raw}T00:00:00Z`);
  return Number.isNaN(parsed.getTime()) ? null : raw;
}

File: backend/schemas/memory.py
"""Core memory / analysis schemas for SnapAct."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


ContentType = Literal[
    "event",
    "quote",
    "knowledge",
    "idea",
    "place",
    "product",
    "job",
    "person_followup",
    "conversation",
    "document",
    "other",
]

IntentMode = Literal["REMEMBER", "EXPLORE", "ACT"]

Urgency = Literal["none", "low", "medium", "high"]

ActionType = Literal[
    "open_url",
    "register",
    "add_calendar",
    "remind",
    "follow_up",
    "view",
    "research",
    "save",
    "none",
]


class Entity(BaseModel):
    name: str
    type: str = "other"


class SuggestedAction(BaseModel):
    type: ActionType
    label: str
    url: str | None = None
    due_at: datetime | None = None
    reason: str | None = None


class TemporalMetadata(BaseModel):
    event_date: str | None = None
    event_end_date: str | None = None
    timezone: str | None = None
    is_upcoming: bool | None = None
    relative_summary: str | None = None


class EventMetadata(BaseModel):
    name: str | None = None
    date: str | None = None
    location: str | None = None
    organizer: str | None = None
    registration_url: str | None = None
    status: str | None = None


class FollowUpMetadata(BaseModel):
    person_name: str | None = None
    topic: str | None = None
    due_hint: str | None = None
    completed: bool = False


class PlaceMetadata(BaseModel):
    name: str | None = None
    address: str | None = None
    city: str | None = None
    category: str | None = None


class ProductMetadata(BaseModel):
    name: str | None = None
    brand: str | None = None
    price_hint: str | None = None
    category: str | None = None


class Citation(BaseModel):
    title: str | None = None
    url: str
    source: Literal["web", "x", "screenshot", "other"] = "web"
    snippet: str | None = None


class AgentActivity(BaseModel):
    """High-level agent activity for UI — never private chain-of-thought."""

    steps: list[str] = Field(default_factory=list)
    web_search_used: bool = False
    x_search_used: bool = False
    live_verification: bool = False
    live_verification_failed: bool = False
    notes: str | None = None


class MemoryAnalysis(BaseModel):
    title: str
    content_type: ContentType
    intent_mode: IntentMode
    intent_summary: str
    description: str
    searchable_text: str
    tags: list[str] = Field(default_factory=list)
    entities: list[Entity] = Field(default_factory=list)
    extracted_text_summary: str | None = None
    actionable: bool = False
    urgency: Urgency = "none"
    needs_live_search: bool = False
    suggested_actions: list[SuggestedAction] = Field(default_factory=list)
    temporal: TemporalMetadata | None = None
    event: EventMetadata | None = None
    person_followup: FollowUpMetadata | None = None
    place: PlaceMetadata | None = None
    product: ProductMetadata | None = None
    confidence: float = 0.5
    answer: str | None = None
    citations: list[Citation] = Field(default_factory=list)
    agent_activity: AgentActivity = Field(default_factory=AgentActivity)
    short_message: str | None = None


class MemoryRecord(BaseModel):
    memory_id: str
    user_id: str
    image_url: str | None = None
    created_at: str
    updated_at: str | None = None
    searchable_text: str = ""
    metadata: dict[str, Any] = Field(default_factory=dict)
    analysis: MemoryAnalysis | None = None
    source: str | None = None
    captured_at: str | None = None
    question: str | None = None
    completed: bool = False
    client_request_id: str | None = None


class MemorySearchHit(BaseModel):
    memory_id: str
    score: float
    image_url: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    analysis: MemoryAnalysis | None = None


class AttentionItem(BaseModel):
    memory_id: str
    title: str
    reason: str
    priority: float
    content_type: ContentType | None = None
    intent_mode: IntentMode | None = None
    image_url: str | None = None
    suggested_action: SuggestedAction | None = None


class HomeFeedPlan(BaseModel):
    generated_at: str
    needs_attention: list[AttentionItem] = Field(default_factory=list)
    upcoming_events: list[AttentionItem] = Field(default_factory=list)
    follow_ups: list[AttentionItem] = Field(default_factory=list)
    suggested_explorations: list[AttentionItem] = Field(default_factory=list)
    quotes: list[AttentionItem] = Field(default_factory=list)
    recent: list[AttentionItem] = Field(default_factory=list)
